Python/ceval.c

```diff
@@ -4518,8 +4519,10 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
             PyTypeObject *cls_type = Py_TYPE(cls);
             assert(cls_type->tp_dictoffset > 0);
             PyObject *dict = *(PyObject **) ((char *)cls + cls_type->tp_dictoffset);
-            // Don't care if no dict -- tp_version_tag should catch anything wrong.
-            DEOPT_IF(dict != NULL && ((PyDictObject *)dict)->ma_keys->dk_version !=
+            // All types should have a dict. dk_version is also an identity check
+            // to ensure `cls` is really a type object. Otherwise, reading
+            // tp_version_tag may segfault.
```

Comment:
Reading `tp_version_tag` cannot segfault if `cls` is a `PyTypeObject`.
However, that isn't checked. Not sure how we missed that.
Can you add `DEOPT_IF(!PyType_Check(cls), LOAD_METHOD);` here.